File: Coding-Interviews/14-I-剪绳子.cpp
#include <climits>
#include <iostream>
#include <vector>
using namespace std;
template<class T>
ostream& operator<<(ostream& x, vector<T> const& y)
{
    x << '[';
    for (auto const& i : y) {
        x << ' ' << i;
    }
    x << " ]";
    return x;
}

class Solution
{
public:
    // int cuttingRope(int n)
    // {
    //     if (n == 1)
    //         return 1;
    //     int                 res = INT_MIN;
    //     vector<vector<int>> dp(n + 1, vector<int>(n + 1, 0));
    //     dp[1][1] = 1;
    //     for (int i = 2; i < n + 1; i++)
    //         dp[i][1] = i;
    //     for (int i = 2; i < n + 1; i++) {
    //         for (int j = 2; j <= i; j++) {
    //             int dp_max = INT_MIN;
    //             for (int k = 1; k <= i - j + 1; k++) {
    //                 dp_max = max(dp_max, dp[i - k][j - 1] * k);
    //             }
    //             dp[i][j] = dp_max;
    //             res      = max(dp[i][j], res);
    //         }
    //     }
    //     return res;
    // }
    int cuttingRope(int n)
    {
        vector<int> dp(n + 1);
        dp[1] = 1;
        for (int i = 2; i < n + 1; i++) {
            int dp_max = INT_MIN;
            for (int j = 1; j < i; j++) {
                dp_max = max(dp_max, max(j * dp[i - j], j * (i - j)));
            }
            dp[i] = dp_max;
        }
        cout << dp;
        return dp[n];
    }
};

int main()
{
    Solution x;
    x.cuttingRope(5);
}

File: Coding-Interviews/14-I-剪绳子.py
# 一维dp, 注意这里的max有三个值, 因为 dp 存的是切成 m 段的最大积, 而 m > 1, 所以至少切两段
# 而计算dp[n]的时候,已经满足了m>1,所以需要比较dp[i-j]和i-j
class Solution:
    def cuttingRope(self, n: int) -> int:
        dp = [0]*(n+1)
        dp[0] = 1
        dp[1] = 1
        for i in range(2, n+1):
            for j in range(1, i):
                dp[i] = max(dp[i], j*dp[i-j], j*(i-j))
        return dp[n]
